src/actions/clipboard.ts

```diff
@@ -0,0 +1,373 @@
+  private pasteCallbackFn(workspace: WorkspaceSvg) {
+    if (!this.copyData || !this.copyWorkspace) return false;
+    const pasteWorkspace = this.copyWorkspace.isFlyout
+      ? workspace
+      : this.copyWorkspace;
+    return this.navigation.paste(this.copyData, pasteWorkspace);
+  }
+
+  /**
+   * Create and register the keyboard shortcut for the cut action.
+   */
+  private registerCutShortcut() {
+    const cutShortcut: ShortcutRegistry.KeyboardShortcut = {
+      name: Constants.SHORTCUT_NAMES.CUT,
+      preconditionFn: this.cutPreconditionFn.bind(this),
+      callback: this.cutCallbackFn.bind(this),
+      keyCodes: [
+        createSerializedKey(KeyCodes.X, [KeyCodes.CTRL]),
+        createSerializedKey(KeyCodes.X, [KeyCodes.ALT]),
+        createSerializedKey(KeyCodes.X, [KeyCodes.META]),
+      ],
+      allowCollision: true,
+    };
+
+    ShortcutRegistry.registry.register(cutShortcut);
+  }
+
+  /**
+   * Register the cut block action as a context menu item on blocks.
+   * This function mixes together the keyboard and context menu preconditions
+   * but only calls the keyboard callback.
+   */
+  private registerCutContextMenuAction() {
+    const cutAction: ContextMenuRegistry.RegistryItem = {
+      displayText: (scope) => {
+        return 'Cut (' + this.getPlatformPrefix() + ' + X)';
+      },
+      preconditionFn: (scope) => {
+        const ws = scope.block?.workspace;
+        if (!ws) return 'hidden';
+
+        return this.cutPreconditionFn(ws) ? 'enabled' : 'disabled';
+      },
+      callback: (scope) => {
+        const ws = scope.block?.workspace;
+        if (!ws) return;
+        return this.cutCallbackFn(ws);
+      },
+      scopeType: ContextMenuRegistry.ScopeType.BLOCK,
+      id: 'blockCutFromContextMenu',
+      weight: this.baseWeight,
+    };
+
+    ContextMenuRegistry.registry.register(cutAction);
+  }
+
+  /**
+   * Callback function for cutting a block from keyboard
+   * navigation. This callback is shared between keyboard shortcuts
+   * and context menu items.
+   *
+   * @param workspace The `WorkspaceSvg` where the shortcut was
+   *     invoked.
+   * @returns True if this function successfully handled cutting.
+   */
+  private cutCallbackFn(workspace: WorkspaceSvg) {
+    const sourceBlock = workspace
+      .getCursor()
+      ?.getCurNode()
+      .getSourceBlock() as BlockSvg;
+    this.copyData = sourceBlock.toCopyData();
+    this.copyWorkspace = sourceBlock.workspace;
+    this.navigation.moveCursorOnBlockDelete(workspace, sourceBlock);
+    sourceBlock.checkAndDelete();
+    return true;
+  }
+
+  /**
+   * Precondition function for cutting a block from keyboard
+   * navigation. This precondition is shared between keyboard shortcuts
+   * and context menu items.
+   *
+   * @param workspace The `WorkspaceSvg` where the shortcut was
+   *     invoked.
+   * @returns True iff `cutCallbackFn` function should be called.
+   */
+  private cutPreconditionFn(workspace: WorkspaceSvg) {
```

Comment:
Probably better to consistently order precondition then callback.

Reply:
Done.